accepted/future-releases/base-interface-final/feature-specification.md

```diff
@@ -0,0 +1,596 @@
+declare any generative constructors.
+
+In practice, mixins are quite different from classes and it's uncommon for users
+to deliberately define a type that is used as both. It's easy to define a class
+without *intending* it to be used as a mixin and then accidentally forbid that
+usage by adding a generative constructor or superclass to the class. That is a
+breaking change to any downstream user that had that class in a `with` clause.
+
+Eventually, we would like classes to not be usable as a mixin by default. Being
+extensible and implementable seems to be the right default for classes based on
+Dart's history and talking to users. But being usable as a mixin by default is
+rarely (but not *never*) useful and often confusing. It makes classes brittle
+for little upside.
+
+Instead, under this proposal we require authors to explicitly opt in to allowing
+the class to be used as a mixin by adding a `mixin` modifier to the class:
+
+```dart
+class OnlyClass {}
+
+class FailUseAsMixin extends OtherSuperclass with OnlyClass {} // Error.
+
+mixin class Both {}
+
+class UsesAsSuperclass extends Both {}
+
+class UsesAsMixin extends OtherSuperclass with Both {} // OK.
+```
+
+This change is guarded by a language version. Any class declaration in a library
+whose language version is before the one this feature ships in can be used as a
+mixin as long as the class meets the mixin restrictions. Classes in libraries
+whose version is later than that must be explicitly marked `mixin class` to
+allow the class to be used in a `with` clause.
+
+When upgrading your library to the new language version, you can preserve the
+existing behavior by adding `mixin` to every class declaration that can be used
+as a mixin. If the class defines a generative constructor or extends anything
+other than `Object`, then it already can't be used as a mixin and no change is
+needed.
+
+## Syntax
+
+This proposal builds on the existing sealed types proposal so the grammar
+includes those changes. The full set of modifiers that can appear before a class
+or mixin are `abstract`, `sealed`, `base`, `interface`, `final`, and `mixin`.
+Some combinations don't make sense:
+
+*   `sealed` implies `abstract`, so they can't be combined.
+*   `sealed` implies non-extensibility, so can't be combined with `interface`
+    or `final`.
+*   `sealed` implies non-implementability, so can't be combined with `base` or
+    `final`.
+*   `base`, `interface`, and `final` all control the same two capabilities so
+    are mutually exclusive.
+*   `mixin` as a modifier can obviously only be applied to a class. It can be
+    combined with any other modifiers that can be applied to a class.
+
+The remaining valid combinations are:
+
+<table><tr><td>
+
+```
+class
+sealed class
+base class
+interface class
+final class
+abstract class
+abstract base class
+abstract interface class
+abstract final class
+```
+
+</td><td>
+
+```
+mixin class
+sealed mixin class
```

Comment:
The restriction stands. You cannot *subtype* a sealed class outside of the library.
Inside the library, you can extend, mix in or implement freely.

It looks a little weird that people outside of the library can even see that it's a `class` or `mixin`, if they can't use it. That goes for a `sealed class` too, but we are just so used to seeing `class` being used as a way to introduce a type (also because we can't avoid exposing an interface today), that we think of it as a neutral type declaration.
If we had `interface Foo { ... }` to produce only a type and an interface, we'd probably use that instead, and use  `sealed interface Foo` to only introduce the type. We probably won't have `sealed type Foo;` by itself.

Reply:
I agree that if we need `mixin` in order to mix things in (your proposed semantics) within a library this is needed.